File: ruby/018.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[][]}
def four_sum(nums, target)
  return [] if nums.length < 4

  nums.sort!

  results = []
  0.upto(nums.length - 4).each do |i|
    (i + 1).upto(nums.length - 3) do |j|
      k = j + 1
      l = nums.length - 1

      while(k < l) do
        current_sum = nums[i] + nums[j] + nums[k] + nums[l]
        if current_sum == target
          results.push([nums[i], nums[j], nums[k], nums[l]])
          current_k_value = nums[k]
          while nums[k + 1] == current_k_value && k < l do
            k += 1
          end
          k += 1
        elsif current_sum > target
          l -= 1
        else
          k += 1
        end
      end
    end
  end

  results.uniq
end

File: javascript/018.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
  nums.sort(function(a, b) { return a - b; }); 

  let result = [];
  const length = nums.length;
  let lastFirstValue = nums[0] - 1;
  for(var i = 0; i < length - 3; ++i) {
    if (nums[i] === lastFirstValue) continue;
    lastFirstValue = nums[i];

    let lastSecondValue = nums[i + 1] - 1;
    for (var j = i + 1; j < length - 2; ++j) {
      if (nums[j] === lastSecondValue) continue;
      lastSecondValue = nums[j];

      const leftTarget = target - nums[i] - nums[j];
      let k = j + 1;
      let l = length - 1;
      while(j < k && k < l && l < length) {
        let currentTarget = nums[k] + nums[l];
        if (currentTarget === leftTarget) {
          result.push([nums[i], nums[j], nums[k], nums[l]]);
          const currentKValue = nums[k];
          while(++k < l && nums[k] === currentKValue){};
        } else if (currentTarget < leftTarget) {
          ++k;
        } else {
          --l;
        }
      }
    }
  }

  return result;
};